README
ABOUT

  Transmission is a fast, easy, and free BitTorrent client.
  It comes in several flavors:

    * A native Mac OS X GUI application
    * GTK+ and Qt GUI applications for Linux, BSD, etc.
    * A headless daemon for servers and routers
    * A web UI for remote controlling any of the above

  Visit https://transmissionbt.com/ for more information.

BUILDING

  Transmission has an Xcode project file (Transmission.xcodeproj)
  for building in Xcode.

  For a more detailed description, and dependancies, visit:
  https://github.com/transmission/transmission/wiki

  Building a Transmission release from the command line:

    $ xz -d -c transmission-2.11.tar.xz | tar xf -
    $ cd transmission-2.11
    $ ./configure
    $ make
    $ sudo make install

  Building Transmission from the nightly builds:

    Download a tarball from https://build.transmissionbt.com/job/trunk-linux-inc/
    and follow the steps from the previous section.

    If you're new to building programs from source code, this is typically 
    easier than building from SVN.

  Building Transmission from SVN (First Time):

    $ svn co svn://svn.transmissionbt.com/Transmission/trunk Transmission
    $ cd Transmission
    $ ./autogen.sh
    $ make
    $ sudo make install

  Building Transmission from SVN (Updating):

    $ cd Transmission
    $ make clean
    $ svn up
    $ ./update-version-h.sh 
    $ make
    $ sudo make install

  Notes for building on Solaris' C compiler:  User av reports success with
  this invocation: ./configure CC=c99 CXX=CC CFLAGS='-D__EXTENSIONS__ -mt'


NEWS
=== Transmission 3.00 (2020/mm/dd) ===
[https://github.com/transmission/transmission/releases/tag/3.00]
=== All Platforms ===
- Allow the RPC server to listen on an IPv6 address (#161)
- Change `TR_CURL_SSL_VERIFY` to `TR_CURL_SSL_NO_VERIFY` and enable verification by default (#334)
- Go back to using hash as base name for resume and torrent files (those stored in configuration directory) (#122)
- Handle "fields" argument in "session-get" RPC request; if "fields" array is present in arguments, only return session fields specified; otherwise return all the fields as before
- Limit the number of incorrect authentication attempts in embedded web server to 100 to prevent brute-force attacks (#371)
- Set idle seed limit range to 1..40320 (4 weeks tops) in all clients (#212)
- Add Peer ID for Xfplay, PicoTorrent, Free Download Manager, Folx, Baidu Netdisk torrent clients (#256, #285, #355, #363, #386)
- Announce `INT64_MAX` as size left if the value is unknown (helps with e.g. Amazon S3 trackers) (#250)
- Add `TCP_FASTOPEN` support (should result in slight speedup) (#184)
- Improve ToS handling on IPv6 connections (#128, #341, #360, #692, #737)
- Abort handshake if establishing DH shared secret fails (leads to crash) (#27)
- Don't switch trackers while announcing (leads to crash) (#297)
- Improve completion scripts execution and error handling; add support for .cmd and .bat files on Windows (#405)
- Maintain a "session ID" file (in temporary directory) to better detect whether session is local or remote; return the ID as part of "session-get" response (TRAC-5348, #861)
- Change torrent location even if no data move is needed (#35)
- Support CIDR-notated blocklists (#230, #741)
- Update the resume file before running scripts (#825)
- Make multiscrape limits adaptive (#837)
- Add labels support to libtransmission and transmission-remote (#822)
- Parse `session-id` header case-insensitively (#765)
- Sanitize suspicious path components instead of rejecting them (#62, #294)
- Load CA certs from system store on Windows / OpenSSL (#446)
- Add support for mbedtls (formely polarssl) and wolfssl (formely cyassl), LibreSSL (#115, #116, #284, #486, #524, #570)
- Fix building against OpenSSL 1.1.0+ (#24)
- Fix quota support for uClibc-ng 1.0.18+ and DragonFly BSD (#42, #58, #312)
- Fix a number of memory leaks (magnet loading, session shutdown, bencoded data parsing) (#56)
- Bump miniupnpc version to 2.0.20170509 (#347)
- CMake-related improvements (Ninja generator, libappindicator, systemd, Solaris and macOS) (#72, #96, #117, #118, #133, #191)
- Switch to submodules to manage (most of) third-party dependencies
- Fail installation on Windows if UCRT is not installed

=== Mac Client ===
- Bump minimum macOS version to 10.10
- Dark Mode support (#644, #722, #757, #779, #788)
- Remove Growl support, notification center is always used (#387)
- Fix autoupdate on High Sierra and up by bumping the Sparkle version (#121, #600)
- Transition to ARC (#336)
- Use proper UTF-8 encoding (with macOS-specific normalization) when setting download/incomplete directory and completion script paths (#11)
- Fix uncaught exception when dragging multiple items between groups (#51)
- Add flat variants of status icons for message log (#134)
- Optimize image resources size (#304, #429)
- Update file icon when file name changes (#37)
- Update translations

=== GTK+ Client ===
- Add queue up/down hotkeys (#158)
- Modernize the .desktop file (#162)
- Add AppData file (#224)
- Add symbolic icon variant for the Gnome top bar and when the high contrast theme is in use (#414, #449)
- Update file icon when its name changes (#37)
- Switch from intltool to gettext for translations (#584, #647)
- Update translations, add new translations for Portuguese (Portugal)

=== Qt Client ===
- Bump minimum Qt version to 5.2
- Fix dropping .torrent files into main window on Windows (#269)
- Fix prepending of drive letter to various user-selected paths on Windows (#236, #307, #404, #437, #699, #723, #877)
- Fix sorting by progress in presence of magnet transfers (#234)
- Fix .torrent file trashing upon addition (#262)
- Add queue up/down hotkeys (#158)
- Reduce torrent properties (file tree) memory usage
- Display tooltips in torrent properties (file tree) in case the names don't fit (#411)
- Improve UI look on hi-dpi displays (YMMV)
- Use session ID (if available) to check if session is local or not (#861)
- Use default (instead of system) locale to be more flexible (#130)
- Modernize the .desktop file (#162)
- Update translations, add new translations for Afrikaans, Catalan, Danish, Greek, Norwegian Bokmål, Slovenian

=== Daemon ===
- Use libsystemd instead of libsystemd-daemon (TRAC-5921)
- Harden transmission-daemon.service by disallowing privileges elevation (#795)
- Fix exit code to be zero when dumping settings (#487)

=== Web Client ===
- Fix tracker error XSS in inspector (CVE-?)
- Fix performance issues due to improper use of `setInterval()` for UI refresh (TRAC-6031)
- Fix recognition of `https://` links in comments field (#41, #180)
- Fix torrent list style in Google Chrome 59+ (#384)
- Show ETA in compact view on non-mobile devices (#146)
- Show upload file button on mobile devices (#320, #431, #956)
- Add keyboard hotkeys for web interface (#351)
- Disable autocompletion in torrent URL field (#367)

=== Utils ===
- Prevent crash in transmission-show displaying torrents with invalid creation date (#609)
- Handle IPv6 RPC addresses in transmission-remote (#247)
- Add `--unsorted` option to transmission-show (#767)
- Widen the torrent-id column in transmission-remote for cleaner formatting (#840)


=== Transmission 2.94 (2018/05/01) ===
[https://github.com/transmission/transmission/releases/tag/2.94 All tickets closed by this release]
==== All Platforms ====
  * Fix building against LibreSSL (#284, #486, #570)
  * Fix building against mbedTLS (#115, #528)
  * Fix torrents ETA calculation (#522)

qt/app.cc
/*
 * This file Copyright (C) 2009-2014 Mnemosyne LLC
 *
 * It may be used under the GNU GPL versions 2 or 3
 * or any future license endorsed by Mnemosyne LLC.
 *
 * $Id$
 */

#include <cassert>
#include <ctime>
#include <iostream>

#include <QDBusConnection>
#include <QDBusConnectionInterface>
#include <QDBusError>
#include <QDBusMessage>
#include <QIcon>
#include <QLibraryInfo>
#include <QMessageBox>
#include <QProcess>
#include <QRect>

#include <libtransmission/transmission.h>
#include <libtransmission/tr-getopt.h>
#include <libtransmission/utils.h>
#include <libtransmission/version.h>

#include "add-data.h"
#include "app.h"
#include "dbus-adaptor.h"
#include "formatter.h"
#include "mainwin.h"
#include "options.h"
#include "prefs.h"
#include "session.h"
#include "session-dialog.h"
#include "torrent-model.h"
#include "watchdir.h"

namespace
{
  const QString DBUS_SERVICE     = QString::fromUtf8 ("com.transmissionbt.Transmission" );
  const QString DBUS_OBJECT_PATH = QString::fromUtf8 ("/com/transmissionbt/Transmission");
  const QString DBUS_INTERFACE   = QString::fromUtf8 ("com.transmissionbt.Transmission" );

  const char * MY_READABLE_NAME ("transmission-qt");

  const tr_option opts[] =
  {
    { 'g', "config-dir", "Where to look for configuration files", "g", 1, "<path>" },
    { 'm', "minimized",  "Start minimized in system tray", "m", 0, NULL },
    { 'p', "port",  "Port to use when connecting to an existing session", "p", 1, "<port>" },
    { 'r', "remote",  "Connect to an existing session at the specified hostname", "r", 1, "<host>" },
    { 'u', "username", "Username to use when connecting to an existing session", "u", 1, "<username>" },
    { 'v', "version", "Show version number and exit", "v", 0, NULL },
    { 'w', "password", "Password to use when connecting to an existing session", "w", 1, "<password>" },
    { 0, NULL, NULL, NULL, 0, NULL }
  };

  const char*
  getUsage ()
  {
    return "Usage:\n"
           "  transmission [OPTIONS...] [torrent files]";
  }

  enum
  {
    STATS_REFRESH_INTERVAL_MSEC   = 3000,
    SESSION_REFRESH_INTERVAL_MSEC = 3000,
    MODEL_REFRESH_INTERVAL_MSEC   = 3000
  };

  bool
  loadTranslation (QTranslator& translator, const QString& name, const QString& localeName,
                   const QStringList& searchDirectories)
  {
    const QString filename = name + QLatin1Char ('_') + localeName;
    for (const QString& directory: searchDirectories)
    {
      if (translator.load (filename, directory))
        return true;
    }

    return false;
  }
}

MyApp::MyApp (int& argc, char ** argv):
  QApplication (argc, argv),
  myPrefs(nullptr),
  mySession(nullptr),
  myModel(nullptr),
  myWindow(nullptr),
  myWatchDir(nullptr),
  myLastFullUpdateTime (0)
{
  const QString MY_CONFIG_NAME = QString::fromUtf8 ("transmission");

  setApplicationName (MY_CONFIG_NAME);

  const QString localeName = QLocale ().name ();

  // install the qt translator
  if (loadTranslation (qtTranslator, QLatin1String ("qt"), localeName, QStringList ()
        << QLibraryInfo::location (QLibraryInfo::TranslationsPath)
#ifdef TRANSLATIONS_DIR
        << QString::fromUtf8 (TRANSLATIONS_DIR)
#endif
        << (applicationDirPath () + QLatin1String ("/translations"))
      ))
    installTranslator (&qtTranslator);

  // install the transmission translator
  if (loadTranslation (appTranslator, MY_CONFIG_NAME, localeName, QStringList ()
#ifdef TRANSLATIONS_DIR
        << QString::fromUtf8 (TRANSLATIONS_DIR)
#endif
        << (applicationDirPath () + QLatin1String ("/translations"))
      ))
    installTranslator (&appTranslator);

  Formatter::initUnits ();

#if defined (_WIN32) || defined (__APPLE__)
  if (QIcon::themeName ().isEmpty ())
    QIcon::setThemeName (QLatin1String ("Faenza"));
#endif

  // set the default icon
  QIcon icon = QIcon::fromTheme (QLatin1String ("transmission"));
  if (icon.isNull ())
    {
      QList<int> sizes;
      sizes << 16 << 22 << 24 << 32 << 48 << 64 << 72 << 96 << 128 << 192 << 256;
      for (const int size: sizes)
        icon.addPixmap (QPixmap (QString::fromLatin1 (":/icons/transmission-%1.png").arg (size)));
    }
  setWindowIcon (icon);

  // parse the command-line arguments
  int c;
  bool minimized = false;
  const char * optarg;
  QString host;
  QString port;
  QString username;
  QString password;
  QString configDir;
  QStringList filenames;
  while ((c = tr_getopt (getUsage(), argc, const_cast<const char**> (argv), opts, &optarg)))
    {
      switch (c)
        {
          case 'g': configDir = QString::fromUtf8 (optarg); break;
          case 'p': port = QString::fromUtf8 (optarg); break;
          case 'r': host = QString::fromUtf8 (optarg); break;
          case 'u': username = QString::fromUtf8 (optarg); break;
          case 'w': password = QString::fromUtf8 (optarg); break;
          case 'm': minimized = true; break;
          case 'v':
            std::cerr << MY_READABLE_NAME << ' ' << LONG_VERSION_STRING << std::endl;
            quitLater ();
            return;
          case TR_OPT_ERR:
            std::cerr << qPrintable(QObject::tr ("Invalid option")) << std::endl;
            tr_getopt_usage (MY_READABLE_NAME, getUsage (), opts);
            quitLater ();
            return;
          default:
            filenames.append (QString::fromUtf8 (optarg));
            break;
        }
    }

  // try to delegate the work to an existing copy of Transmission
  // before starting ourselves...
  QDBusConnection bus = QDBusConnection::sessionBus ();
  if (bus.isConnected ())
  {
    bool delegated = false;
    for (const QString& filename: filenames)
      {
        QDBusMessage request = QDBusMessage::createMethodCall (DBUS_SERVICE,
                                                               DBUS_OBJECT_PATH,
                                                               DBUS_INTERFACE,
                                                               QString::fromUtf8 ("AddMetainfo"));
        QList<QVariant> arguments;
        AddData a (filename);
        switch (a.type)
          {
            case AddData::URL:      arguments.push_back (a.url.toString ()); break;
            case AddData::MAGNET:   arguments.push_back (a.magnet); break;
            case AddData::FILENAME: arguments.push_back (QString::fromLatin1 (a.toBase64 ())); break;
            case AddData::METAINFO: arguments.push_back (QString::fromLatin1 (a.toBase64 ())); break;
            default:                break;
          }
        request.setArguments (arguments);

        QDBusMessage response = bus.call (request);
        //std::cerr << qPrintable (response.errorName ()) << std::endl;
        //std::cerr << qPrintable (response.errorMessage ()) << std::endl;
        arguments = response.arguments ();
        delegated |= (arguments.size ()==1) && arguments[0].toBool ();
      }

    if (delegated)
      {
        quitLater ();
        return;
      }
  }

  // set the fallback config dir
  if (configDir.isNull ())
    configDir = QString::fromUtf8 (tr_getDefaultConfigDir ("transmission"));

  // ensure our config directory exists
  QDir dir (configDir);
  if (!dir.exists ())
    dir.mkpath (configDir);

  // is this the first time we've run transmission?
  const bool firstTime = !dir.exists (QLatin1String ("settings.json"));

  // initialize the prefs
  myPrefs = new Prefs (configDir);
  if (!host.isNull ())
    myPrefs->set (Prefs::SESSION_REMOTE_HOST, host);
  if (!port.isNull ())
    myPrefs->set (Prefs::SESSION_REMOTE_PORT, port.toUInt ());
  if (!username.isNull ())
    myPrefs->set (Prefs::SESSION_REMOTE_USERNAME, username);
  if (!password.isNull ())
    myPrefs->set (Prefs::SESSION_REMOTE_PASSWORD, password);
  if (!host.isNull () || !port.isNull () || !username.isNull () || !password.isNull ())
    myPrefs->set (Prefs::SESSION_IS_REMOTE, true);
  if (myPrefs->getBool (Prefs::START_MINIMIZED))
    minimized = true;

  // start as minimized only if the system tray present
  if (!myPrefs->getBool (Prefs::SHOW_TRAY_ICON))
    minimized = false;

  mySession = new Session (configDir, *myPrefs);
  myModel = new TorrentModel (*myPrefs);
  myWindow = new TrMainWindow (*mySession, *myPrefs, *myModel, minimized);
  myWatchDir = new WatchDir (*myModel);

  // when the session gets torrent info, update the model
  connect (mySession, SIGNAL (torrentsUpdated (tr_variant*,bool)), myModel, SLOT (updateTorrents (tr_variant*,bool)));
  connect (mySession, SIGNAL (torrentsUpdated (tr_variant*,bool)), myWindow, SLOT (refreshActionSensitivity ()));
  connect (mySession, SIGNAL (torrentsRemoved (tr_variant*)), myModel, SLOT (removeTorrents (tr_variant*)));
  // when the session source gets changed, request a full refresh
  connect (mySession, SIGNAL (sourceChanged ()), this, SLOT (onSessionSourceChanged ()));
  // when the model sees a torrent for the first time, ask the session for full info on it
  connect (myModel, SIGNAL (torrentsAdded (QSet<int>)), mySession, SLOT (initTorrents (QSet<int>)));
  connect (myModel, SIGNAL (torrentsAdded (QSet<int>)), this, SLOT (onTorrentsAdded (QSet<int>)));

  mySession->initTorrents ();
  mySession->refreshSessionStats ();

  // when torrents are added to the watch directory, tell the session
  connect (myWatchDir, SIGNAL (torrentFileAdded (QString)), this, SLOT (addTorrent (QString)));

  // init from preferences
  QList<int> initKeys;
  initKeys << Prefs::DIR_WATCH;
  for (const int key: initKeys)
    refreshPref (key);
  connect (myPrefs, SIGNAL (changed (int)), this, SLOT (refreshPref (const int)));

  QTimer * timer = &myModelTimer;
  connect (timer, SIGNAL (timeout ()), this, SLOT (refreshTorrents ()));
  timer->setSingleShot (false);
  timer->setInterval (MODEL_REFRESH_INTERVAL_MSEC);
  timer->start ();

  timer = &myStatsTimer;
  connect (timer, SIGNAL (timeout ()), mySession, SLOT (refreshSessionStats ()));
  timer->setSingleShot (false);
  timer->setInterval (STATS_REFRESH_INTERVAL_MSEC);
  timer->start ();

  timer = &mySessionTimer;
  connect (timer, SIGNAL (timeout ()), mySession, SLOT (refreshSessionInfo ()));
  timer->setSingleShot (false);
  timer->setInterval (SESSION_REFRESH_INTERVAL_MSEC);
  timer->start ();

  maybeUpdateBlocklist ();

  if (!firstTime)
    {
      mySession->restart ();
    }
  else
    {
      QDialog * d = new SessionDialog (*mySession, *myPrefs, myWindow);
      d->show ();
    }

  if (!myPrefs->getBool (Prefs::USER_HAS_GIVEN_INFORMED_CONSENT))
    {
      QMessageBox * dialog = new QMessageBox (QMessageBox::Information, QString (),
                                              tr ("<b>Transmission is a file sharing program.</b>"),
                                              QMessageBox::Ok | QMessageBox::Cancel, myWindow);
      dialog->setInformativeText (tr ("When you run a torrent, its data will be made available to others by means of upload. "
                                      "Any content you share is your sole responsibility."));
      dialog->button (QMessageBox::Ok)->setText (tr ("I &Agree"));
      dialog->setDefaultButton (QMessageBox::Ok);
      dialog->setModal (true);

      connect (dialog, SIGNAL (finished (int)), this, SLOT (consentGiven (int)));

      dialog->setAttribute (Qt::WA_DeleteOnClose);
      dialog->show ();
    }

  for (const QString& filename: filenames)
    addTorrent (filename);

  // register as the dbus handler for Transmission
  if (bus.isConnected ())
    {
      new TrDBusAdaptor (this);
      if (!bus.registerService (DBUS_SERVICE))
        std::cerr << "couldn't register " << qPrintable (DBUS_SERVICE) << std::endl;
      if (!bus.registerObject (DBUS_OBJECT_PATH, this))
        std::cerr << "couldn't register " << qPrintable (DBUS_OBJECT_PATH) << std::endl;
    }
}

void
MyApp::quitLater ()
{
  QTimer::singleShot (0, this, SLOT (quit ()));
}

/* these functions are for popping up desktop notifications */

void
MyApp::onTorrentsAdded (const QSet<int>& torrents)
{
  if (!myPrefs->getBool (Prefs::SHOW_NOTIFICATION_ON_ADD))
    return;

  for (const int id: torrents)
    {
      Torrent * tor = myModel->getTorrentFromId (id);

      if (tor->name ().isEmpty ()) // wait until the torrent's INFO fields are loaded
        {
          connect (tor, SIGNAL (torrentChanged (int)), this, SLOT (onNewTorrentChanged (int)));
        }
      else
        {
          onNewTorrentChanged (id);

          if (!tor->isSeed ())
            connect (tor, SIGNAL (torrentCompleted (int)), this, SLOT (onTorrentCompleted (int)));
        }
    }
}

void
MyApp::onTorrentCompleted (int id)
{
  Torrent * tor = myModel->getTorrentFromId (id);

  if (tor)
    {
      if (myPrefs->getBool (Prefs::SHOW_NOTIFICATION_ON_COMPLETE))
        notifyApp (tr ("Torrent Completed"), tor->name ());

      if (myPrefs->getBool (Prefs::COMPLETE_SOUND_ENABLED))
        {
#if defined (Q_OS_WIN) || defined (Q_OS_MAC)
          beep ();
#else
          QProcess::execute (myPrefs->getString (Prefs::COMPLETE_SOUND_COMMAND));
#endif
        }

      disconnect (tor, SIGNAL (torrentCompleted (int)), this, SLOT (onTorrentCompleted (int)));
    }
}

void
MyApp::onNewTorrentChanged (int id)
{
  Torrent * tor = myModel->getTorrentFromId (id);

  if (tor && !tor->name ().isEmpty ())
    {
      const int age_secs = tor->dateAdded ().secsTo (QDateTime::currentDateTime ());
      if (age_secs < 30)
        notifyApp (tr ("Torrent Added"), tor->name ());

      disconnect (tor, SIGNAL (torrentChanged (int)), this, SLOT (onNewTorrentChanged (int)));

      if (!tor->isSeed ())
        connect (tor, SIGNAL (torrentCompleted (int)), this, SLOT (onTorrentCompleted (int)));
    }
}

/***
****
***/

void
MyApp::consentGiven (int result)
{
  if (result == QMessageBox::Ok)
    myPrefs->set<bool> (Prefs::USER_HAS_GIVEN_INFORMED_CONSENT, true);
  else
    quit ();
}

MyApp::~MyApp ()
{
  if (myPrefs != nullptr && myWindow != nullptr)
    {
      const QRect mainwinRect (myWindow->geometry ());
      myPrefs->set (Prefs::MAIN_WINDOW_HEIGHT, std::max (100, mainwinRect.height ()));
      myPrefs->set (Prefs::MAIN_WINDOW_WIDTH, std::max (100, mainwinRect.width ()));
      myPrefs->set (Prefs::MAIN_WINDOW_X, mainwinRect.x ());
      myPrefs->set (Prefs::MAIN_WINDOW_Y, mainwinRect.y ());
    }

  delete myWatchDir;
  delete myWindow;
  delete myModel;
  delete mySession;
  delete myPrefs;
}

/***
****
***/

void
MyApp::refreshPref (int key)
{
  switch (key)
    {
      case Prefs::BLOCKLIST_UPDATES_ENABLED:
        maybeUpdateBlocklist ();
        break;

      case Prefs::DIR_WATCH:
      case Prefs::DIR_WATCH_ENABLED:
        {
          const QString path (myPrefs->getString (Prefs::DIR_WATCH));
          const bool isEnabled (myPrefs->getBool (Prefs::DIR_WATCH_ENABLED));
          myWatchDir->setPath (path, isEnabled);
          break;
        }

      default:
        break;
    }
}

void
MyApp::maybeUpdateBlocklist ()
{
  if (!myPrefs->getBool (Prefs::BLOCKLIST_UPDATES_ENABLED))
    return;

  const QDateTime lastUpdatedAt = myPrefs->getDateTime (Prefs::BLOCKLIST_DATE);
  const QDateTime nextUpdateAt = lastUpdatedAt.addDays (7);
  const QDateTime now = QDateTime::currentDateTime ();

  if (now < nextUpdateAt)
    {
      mySession->updateBlocklist ();
      myPrefs->set (Prefs::BLOCKLIST_DATE, now);
    }
}

void
MyApp::onSessionSourceChanged ()
{
  mySession->initTorrents ();
  mySession->refreshSessionStats ();
  mySession->refreshSessionInfo ();
}

void
MyApp::refreshTorrents ()
{
  // usually we just poll the torrents that have shown recent activity,
  // but we also periodically ask for updates on the others to ensure
  // nothing's falling through the cracks.
  const time_t now = time (NULL);
  if (myLastFullUpdateTime + 60 >= now)
    {
      mySession->refreshActiveTorrents ();
    }
  else
    {
      myLastFullUpdateTime = now;
      mySession->refreshAllTorrents ();
    }
}

/***
****
***/

void
MyApp::addTorrent (const QString& key)
{
  const AddData addme (key);

  if (addme.type != addme.NONE)
    addTorrent (addme);
}

void
MyApp::addTorrent (const AddData& addme)
{
  if (!myPrefs->getBool (Prefs::OPTIONS_PROMPT))
    {
      mySession->addTorrent (addme);
    }
  else
    {
      OptionsDialog * o = new OptionsDialog (*mySession, *myPrefs, addme, myWindow);
      o->show ();
    }

  raise ();
}

/***
****
***/

void
MyApp::raise ()
{
  alert (myWindow);
}

bool
MyApp::notifyApp (const QString& title, const QString& body) const
{
  const QString dbusServiceName   = QString::fromUtf8 ("org.freedesktop.Notifications");
  const QString dbusInterfaceName = QString::fromUtf8 ("org.freedesktop.Notifications");
  const QString dbusPath          = QString::fromUtf8 ("/org/freedesktop/Notifications");

  QDBusMessage m = QDBusMessage::createMethodCall (dbusServiceName, dbusPath, dbusInterfaceName, QString::fromUtf8 ("Notify"));
  QList<QVariant> args;
  args.append (QString::fromUtf8 ("Transmission")); // app_name
  args.append (0U);                                   // replaces_id
  args.append (QString::fromUtf8 ("transmission")); // icon
  args.append (title);                                // summary
  args.append (body);                                 // body
  args.append (QStringList ());                       // actions - unused for plain passive popups
  args.append (QVariantMap ());                       // hints - unused atm
  args.append (static_cast<int32_t> (-1));            // use the default timeout period
  m.setArguments (args);
  QDBusMessage replyMsg = QDBusConnection::sessionBus ().call (m);
  //std::cerr << qPrintable (replyMsg.errorName ()) << std::endl;
  //std::cerr << qPrintable (replyMsg.errorMessage ()) << std::endl;
  return (replyMsg.type () == QDBusMessage::ReplyMessage) && !replyMsg.arguments ().isEmpty ();
}

/***
****
***/

int
tr_main (int    argc,
         char * argv[])
{
  MyApp app (argc, argv);
  return app.exec ();
}
